### DISCUSSION TITLE
Variations for Headings

### DISCUSSION THREAD

**Discussion by pkazmier** [MeanderingProgrammer/render-markdown.nvim#273] [Show and tell]
I spent more time than I'd like to admit this morning trying different variations for headings, so I thought I'd share some in case others would benefit. Each section has a Details expansion that contains the configuration used for the headings:

---
<details><summary>Configuration</summary>
<p>

```lua
{
    heading = {
      sign = false,
      border = true,
      below = "▔",
      above = "▁",
      left_pad = 0,
      position = "left",
      icons = {
        "█ ",
        "██ ",
        "███ ",
        "████ ",
        "█████ ",
        "██████ ",
      },
    },
  }
```

</p>
</details> 
<img width="763" alt="image" src="https://github.com/user-attachments/assets/fa8c0148-6703-4916-b005-a8462f3ddc2e" />

---
<details><summary>Configuration</summary>
<p>

```lua
{
    heading = {
      sign = false,
      border = true,
      below = "▔",
      above = "▁",
      left_pad = 0,
      position = "left",
      icons = {
        "█ ",
        "██ ",
        "███ ",
        "████ ",
        "█████ ",
        "██████ ",
      },
    },
  }
```

</p>
</details> 
<img width="763" alt="image" src="https://github.com/user-attachments/assets/bb6a5b1a-1950-41cb-92c0-5e6b76b3bf42" />

---
<details><summary>Configuration</summary>
<p>

```lua
{
    heading = {
      sign = false,
      border = true,
      below = "▔",
      above = "▁",
      left_pad = 0,
      position = "left",
      icons = {
        " ",
        " ",
        " ",
        " ",
        " ",
        " ",
      },
    },
  }
```

</p>
</details> 
<img width="763" alt="image" src="https://github.com/user-attachments/assets/ab768116-6f41-4436-bcb8-a3d50aae4960" />

---
<details><summary>Configuration</summary>
<p>

```lua
{
    heading = {
      sign = false,
      border = true,
      width = "block",
      below = "▔",
      above = "▁",
      left_pad = 0,
      right_pad = 4,
      position = "left",
      icons = {
        "█ ",
        "██ ",
        "███ ",
        "████ ",
        "█████ ",
        "██████ ",
      },
    },
  }
```

</p>
</details> 
<img width="763" alt="image" src="https://github.com/user-attachments/assets/00050410-78f9-46fc-984e-e0df38a659e4" />

---
<details><summary>Configuration</summary>
<p>

```lua
{
    heading = {
      sign = false,
      border = true,
      width = "block",
      below = "▔",
      above = "▁",
      left_pad = 0,
      right_pad = 4,
      position = "left",
      icons = {
        "█ ",
        "██ ",
        "███ ",
        "████ ",
        "█████ ",
        "██████ ",
      },
    },
  }
```

</p>
</details> 
<img width="763" alt="image" src="https://github.com/user-attachments/assets/3024f5bb-2f84-4ba5-b8ed-f69c4394002d" />

---
<details><summary>Configuration</summary>
<p>

```lua
{
    heading = {
      sign = false,
      border = true,
      width = "block",
      below = "▔",
      above = "▁",
      left_pad = 0,
      right_pad = 4,
      position = "left",
      icons = {
        " ",
        " ",
        " ",
        " ",
        " ",
        " ",
      },
    },
  }
```

</p>
</details> 
<img width="763" alt="image" src="https://github.com/user-attachments/assets/f2b25a9c-b291-42a0-b779-135c1b4afa93" />

---
<details><summary>Configuration</summary>
<p>

```lua
{
    heading = {
      backgrounds = {},
      sign = false,
      border = true,
      below = "▔",
      above = "▁",
      left_pad = 0,
      position = "left",
      icons = {
        "█ ",
        "██ ",
        "███ ",
        "████ ",
        "█████ ",
        "██████ ",
      },
    },
  }
```

</p>
</details> 
<img width="763" alt="image" src="https://github.com/user-attachments/assets/0ff5a7ed-9dae-4d16-8e21-65fa1a8aa9e5" />

---
<details><summary>Configuration</summary>
<p>

```lua
{
    heading = {
      backgrounds = {},
      sign = false,
      border = true,
      below = "▔",
      above = "▁",
      left_pad = 0,
      position = "left",
      icons = {
        "█ ",
        "██ ",
        "███ ",
        "████ ",
        "█████ ",
        "██████ ",
      },
    },
  }
```

</p>
</details> 
<img width="763" alt="image" src="https://github.com/user-attachments/assets/275ed870-bcb8-4ac8-83a4-4d9398d55274" />

---
<details><summary>Configuration</summary>
<p>

```lua
{
    heading = {
      backgrounds = {},
      sign = false,
      border = true,
      below = "▔",
      above = "▁",
      left_pad = 0,
      position = "left",
      icons = {
        " ",
        " ",
        " ",
        " ",
        " ",
        " ",
      },
    },
  }
```

</p>
</details> 
<img width="763" alt="image" src="https://github.com/user-attachments/assets/2cd45c23-1bf5-4685-826c-06795c90bf8d" />

**Comment by MeanderingProgrammer** [comment]
Hi, thank you for sharing! I hope the tinkering was fun, I've definitely gotten sucked into getting a few plugins to work just the way I want them to haha.

Would you able to put the config you used next to the associated screenshot? Just so it's easier for people to copy paste it and get the same result.

**Reply by pkazmier** [reply]
Sure. I also added several more variants.

**Comment by sgon00** [comment]
Thanks a lot for your sharing. This is my variation:

![render-markdown-sample01](https://github.com/user-attachments/assets/c3bfe140-6b42-48a5-84e0-eb4da8ad594b)

### Configuration

I'm using [NvChad](https://nvchad.com/) config. Here are some snippets from my config files that follow the NvChad template.

* Theme (`~/.config/nvim/lua/chadrc.lua`):

```lua
M.base46 = {
  theme = "nord",
```

* render-markdown.nvim Config (`~/.config/nvim/lua/plugins/init.lua`)

```lua
local highlight_md_heading = "MyHighlightHeading"
vim.api.nvim_set_hl(0, highlight_md_heading, { bg = "#3e4452", fg = "#81a1c1" })

My_Render_Markdown_Opts = {
  render_modes = true, -- enable all modes
  heading = {
    width = {'full', 'block'},
    border = true,
    border_virtual = true,
    border_prefix = true,
    min_width = 40,
    sign = false,
    icons = {
      "█ ",
      "██ ",
      "███ ",
      "████ ",
      "█████ ",
      "██████ ",
    },
    backgrounds = { highlight_md_heading },
  },
  code = {
    width = 'block',
    left_pad = 2,
    right_pad = 2,
    sign = false,
    left_margin = 2,
    position = 'right', -- language indicator position
  },
}

return {
  {
    'MeanderingProgrammer/render-markdown.nvim',
    ft = { 'markdown', 'quarto' },
    dependencies = { 'nvim-treesitter/nvim-treesitter', 'echasnovski/mini.nvim' }, -- if you use the mini.nvim suite
    ---@module 'render-markdown'
    ---@type render.md.UserConfig
    opts = My_Render_Markdown_Opts,
  },
```

**Reply by pkazmier** [reply]
Looks great. I like using the thicker headlines too, but the problem I have is that I start my L1 headings on the first line of the buffer. And, unfortunately, the top part of the thicker border is not shown, so it has the effect of making the L1 header visually less prominent / important than the rest. That's why I ultimately settled on the middle ground of the slightly thicker border as it does not make it as noticeable. I wish there was a way to have vim insert some visual space above line 1. I'm sure it can be done as sometimes my window will get in a state where the full thickness of that first line is shown. I suppose I could do what you did here and just start on the second line, but my formatter is just going to autocorrect that, plus my brain wants my L1 to start on line 1 anyways! In any case, your headlines look nice without Nord setup. Nice work!

**Reply by sgon00** [reply]
Yeah, after using neovim + nvchad + render-markdown, I change all my markdown files to start on the second line. 😀
If you press `PageUp`, you can see the visual space above line 1. But I still don't like that workaround by always pressing `PageUp` after opening the file.

**Reply by pkazmier** [reply]
Ah, thanks for the workaround!

**Reply by pkazmier** [reply]
Oh, could you post your configuration for others that don't know about `min_width`? The maintainer asked me to go post all my configs next to each one as well.

**Reply by pkazmier** [reply]
> But I still don't like that workaround by always pressing `PageUp` after opening the file.

@sgon00 I think I found a workaround. It's ugly, but seems to work for me. You might want to give it a try. I documented it in issue https://github.com/MeanderingProgrammer/render-markdown.nvim/issues/123#issuecomment-2571655526

**Reply by malasee** [reply]
@sgon00 This looks great!  Could you please share your config for this?

**Reply by sgon00** [reply]
@malasee It's great to hear you like it. I've updated my comment to include the detailed configuration.

**Reply by malasee** [reply]
Thanks a lot for sharing :)

**Reply by MeanderingProgrammer** [reply]
Thanks for sharing! 

I've updated your configuration to only define a single highlight group, there's no need to define 6 if the values are the same.

**Reply by sgon00** [reply]
Sure, thanks for the update. I initially used 6 values just in case I needed to change some heading colors in the future, but it seems one color is fine for me. Too many colors would be distracting.

By the way, I forgot to mention I'm using NvChad config; I've updated that information.

**Reply by pkazmier** [reply]
I made a variation of your variation 😄 
<img width="739" alt="image" src="https://github.com/user-attachments/assets/c67cea0f-7999-473b-adc1-0f28b386cd15" />

```lua
    heading = {
      backgrounds = { "DiffChange" },
      sign = false,
      border = true,
      border_prefix = true,
      left_pad = 0,
      right_pad = 4,
      position = "overlay",
      icons = {
        " ",
        " ",
        " ",
        " ",
        " ",
        " ",
      },

```

**Comment by rodhash** [comment]
> <img alt="image" width="763" src="https://private-user-images.githubusercontent.com/747855/400093730-fa8c0148-6703-4916-b005-a8462f3ddc2e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI3MTI5NjgsIm5iZiI6MTc0MjcxMjY2OCwicGF0aCI6Ii83NDc4NTUvNDAwMDkzNzMwLWZhOGMwMTQ4LTY3MDMtNDkxNi1iMDA1LWE4NDYyZjNkZGMyZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyM1QwNjUxMDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jYzQyM2UwYWRkNDBiNjA3N2U3NGFhYjBhY2U1NDE1YmMxNzMxZWZjOWRlMGE3ZjhiM2YwZjQ3MGJiODRhODQ2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.vjRF9RvxvxpCtY9UIerfdv1VxtUcw0gjRYFJWxQBe48">

Hey @pkazmier , thanks for sharing!

That's a nice font, is this a nerd font? Which one is this pls?

**Reply by pkazmier** [reply]
It’s [Maple Mono](https://github.com/subframe7536/maple-font). It’s a lovely font.

**Reply by rodhash** [reply]
> It’s [Maple Mono](https://github.com/subframe7536/maple-font). It’s a lovely font.

Thank you!

**Comment by pkazmier** [comment]
Here is my latest variation that I've been using:
<img width="1361" height="1409" alt="image" src="https://github.com/user-attachments/assets/0dbc59a8-2975-4130-8156-5742f06b5efc" />


And my configuration:
```lua
    heading = {
      width = "block",
      backgrounds = {
        "MiniStatusLineModeNormal",
        "MiniStatusLineModeInsert",
        "MiniStatusLineModeReplace",
        "MiniStatusLineModeVisual",
        "MiniStatusLineModeCommand",
        "MiniStatusLineModeOther",
      },
      sign = true,
      left_pad = 1,
      right_pad = 0,
      position = "right",
      icons = {
        "",
        "",
        "",
        "",
        "",
        "",
      },
    },
```

Importantly, the `backgrounds` highlight groups `fg` color is the `bg` color of `Normal`.
